### PR TITLE
Add elasticsearch instrumentation

### PR DESCRIPTION
**Pull request by SimunKaracic**
Add elasticsearch instrumentation for the official java High and Low level REST client (current version, 7.9). 

Simple implementation, records a span for each sync and async request with 2 tags (endpoint, method), 
and creates histogram of request sizes.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -319,6 +320,22 @@ lazy val `kamon-cassandra` = (project in file("instrumentation/kamon-cassandra")
...
+  .settings(
+    fork in (Test,run) := true,
+    libraryDependencies ++= Seq(
+      kanelaAgent % "provided",
+      "org.elasticsearch.client" % "elasticsearch-rest-client" % "7.9.1" % "provided",
+      "org.elasticsearch.client" % "elasticsearch-rest-high-level-client" % "7.9.1" % "provided",
+      scalatest % "test",
+      logbackClassic % "test",
+      "com.dimafeng" % "testcontainers-scala_2.12" % "0.38.3" % "test",
+      "com.dimafeng" % "testcontainers-scala-elasticsearch_2.12" % "0.38.3" % "test"
```
Comment (ivantopo) [inline build.sbt:336]: Why is the scala version included in these two artifacts? That will probably break cross testing.

File: instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/highlevel/SyncClientInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package kamon.instrumentation.elasticsearch.highlevel;
+
+import kanela.agent.libs.net.bytebuddy.asm.Advice;
+
+public class SyncClientInstrumentation {
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/highlevel/SyncClientInstrumentation.java:5]: Is there any particular reason why `AsyncClientInstrumentation` and `SyncClientInstrumentation` aren't just the same advice? Also, a more specific name would help :)

File: instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/ESInstrumentation.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  }
+}
+
+object RequestSizeHistogram {
+  private val histogram = Kamon.histogram("elastic.request.size").withoutTags()
+
+  def record(entity: HttpEntity): Unit = {
+    Option(entity)
+      .map(_.getContentLength)
+      .filter(_ >= 0)
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/ESInstrumentation.scala:38]: In what cases could the content be negative or zero?

File: instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/RequestNameConverter.scala [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package kamon.instrumentation.elasticsearch
+
+object RequestNameConverter {
+  def convert(className: String, fallback: String): String = {
+    // converts class name to something more appropriate
+    // e.g. ListTasksRequest to list.tasks.requests
+    Option(className)
+      .getOrElse(fallback)
+      .replaceAll("([^_A-Z])([A-Z])", "$1.$2")
+      .toLowerCase
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/RequestNameConverter.scala:10]: Few things that would be better here:
  - Add the `elasticsearch` prefix to all names.
  - Drop the `Request` at the end of the name.
  - Cache these transformed operation names. The number of possible options is bound so it wouldn't be a problem.

File: build.sbt [hunk 5]
```diff
@@ -319,6 +320,22 @@ lazy val `kamon-cassandra` = (project in file("instrumentation/kamon-cassandra")
...
+    )
+  ).dependsOn(`kamon-core`, `kamon-instrumentation-common`, `kamon-testkit` % "test", `kamon-executors`)
```
Comment (ivantopo) [inline build.sbt:338]: Why the dependency on executors?

File: instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncRestClientInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+public class AsyncRestClientInstrumentation {
+
+    @Advice.OnMethodEnter
+    public static void enter(
+            @Advice.Argument(0) Request request,
+            @Advice.Argument(value = 1, readOnly = false) ResponseListener responseListener) {
+        final String operationName =
+                RequestNameConverter.convert(
+                        AsyncClientInstrumentation.requestClassName.get(),
+                        "elastic.async.request");
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncRestClientInstrumentation.java:19]: Please change this and all other mentions of `elastic` to `elasticsearch`

**Review by ivantopo — CHANGES_REQUESTED** [review]

File: instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/ESInstrumentation.scala (hunk 3 above)
Comment (SimunKaracic) [reply instrumentation/kamon-elasticsearch/src/main/scala/kamon/instrumentation/elasticsearch/ESInstrumentation.scala:38]: Just being cautious. From the docs:

```
     * @return  the number of bytes of the content, or
     *          a negative number if unknown. If the content length is known
     *          but exceeds {@link java.lang.Long#MAX_VALUE Long.MAX_VALUE},
     *          a negative number is returned.
```

File: instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/highlevel/SyncClientInstrumentation.java (hunk 2 above)
Comment (SimunKaracic) [reply instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/highlevel/SyncClientInstrumentation.java:5]: D'oh, not reason. 
Will merge them with a prettier name.

File: build.sbt (hunk 5 above)
Comment (SimunKaracic) [reply build.sbt:338]: Copy paste artifact, removing...

File: build.sbt (hunk 1 above)
Comment (SimunKaracic) [reply build.sbt:336]: Yeah after some trying out, I'll include specific version for each of those libraries, depending on the version of scala used, e.g.
```
{
        if(scalaBinaryVersion.value == "2.11") {
          "com.dimafeng" % "testcontainers-scala-elasticsearch_2.11" % "0.38.3" % "test"
        } else if(scalaBinaryVersion.value == "2.12") {
          "com.dimafeng" % "testcontainers-scala-elasticsearch_2.12" % "0.38.3" % "test"
        } else {
          "com.dimafeng" % "testcontainers-scala-elasticsearch_2.13" % "0.38.3" % "test"
        }
      }
```

**Comment by SimunKaracic** [comment]
`kamon-mongo` tests failing...

File: build.sbt [hunk 7]
```diff
@@ -319,6 +320,38 @@ lazy val `kamon-cassandra` = (project in file("instrumentation/kamon-cassandra")
...
+    fork in (Test,run) := true,
+    libraryDependencies ++= Seq(
+      kanelaAgent % "provided",
+      "org.elasticsearch.client" % "elasticsearch-rest-client" % "7.9.1" % "provided",
+      "org.elasticsearch.client" % "elasticsearch-rest-high-level-client" % "7.9.1" % "provided",
+      scalatest % "test",
+      logbackClassic % "test",
+      {
+        if(scalaBinaryVersion.value == "2.11") {
+          "com.dimafeng" % "testcontainers-scala_2.11" % "0.38.3" % "test"
```
Comment (ElfoLiNk) [inline build.sbt:337]: Just use %% `"com.dimafeng" %% "testcontainers-scala"`

Comment (SimunKaracic) [reply build.sbt:337]: I swear I tried that and it didn't work... I must have forgotten for use `sbt reload` or something. 

Thx for the feedback :+1:

File: instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncElasticsearchRestClientInstrumentation.java [hunk 8]
```diff
@@ -15,7 +15,7 @@ public static void enter(
         final String operationName =
                 RequestNameConverter.convert(
                         HighLevelElasticsearchClientInstrumentation.requestClassName.get(),
-                        "elasticsearch.async");
+                        "AsyncRequest");
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncElasticsearchRestClientInstrumentation.java:18]: This one should be `elasticsearch/AsyncRequest`, same goes for `elasticsearch/SyncRequest`.

File: instrumentation/kamon-elasticsearch/src/main/resources/reference.conf [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    elasticsearch-driver {
+
+      name = "Elasticsearch client"
+      description = "Elasticsearch Client Instrumentation"
+      instrumentations = [
+        "kamon.instrumentation.elasticsearch.ESInstrumentation"
+      ]
+
+      within = [
+        "org.elasticsearch..*"
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/resources/reference.conf:19]: Can this be narrowed down?

File: instrumentation/kamon-elasticsearch/src/main/resources/reference.conf [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+kamon.instrumentation.elasticsearch {
+}
+
+kanela {
+  modules {
+    elasticsearch-driver {
+
+      name = "Elasticsearch client"
+      description = "Elasticsearch Client Instrumentation"
```
Comment (ivantopo) [inline instrumentation/kamon-elasticsearch/src/main/resources/reference.conf:13]: Name: Elasticsearch Client
Description: Provides tracing of client calls made with the official Elasticsearch Client library.

**Review by ivantopo — CHANGES_REQUESTED** [review]

File: instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncElasticsearchRestClientInstrumentation.java (hunk 8 above)
Comment (SimunKaracic) [reply instrumentation/kamon-elasticsearch/src/main/java/kamon/instrumentation/elasticsearch/AsyncElasticsearchRestClientInstrumentation.java:18]: "AsyncRequest" and "SyncRequest" are fallbacks for classnames, like "ListTasksRequest" (which might be null).
So they resulting operation name is "elasticsearch/ClassName".

File: instrumentation/kamon-elasticsearch/src/main/resources/reference.conf (hunk 9 above)
Comment (SimunKaracic) [reply instrumentation/kamon-elasticsearch/src/main/resources/reference.conf:19]: Narrowed to "org.elasticseearch.client..*"

**Pull request merged**